### PR TITLE
[#9673] Fix feedback session names with '+' encoding

### LINKED ISSUE #9673: Feedback sessions with + in their name get replaced by ' ' when requested
- **Environment**: `master` branch at commit a384354b8ecf3cb87006184185724f16560cf14b

**Steps to reproduce**
1. Create a new feedback session with `+` in the name, i.e. `Super+++Session`
2. Attempt to view number of responses in that session

**Expected behaviour**
Number of responses is shown as `0 / X`

**Actual behaviour**
The loading part appears and loops forever while an error message appears "Feedback session does not exist"
image

**Additional info**

This appears to be due to the request not sending the `+` character but instead replacing it with ` ` (space) in it's place.
This also causes the feedback session to be uneditable as no questions will load in the edit page.

### PR DESCRIPTION
**Pull request by madanalogy**
Fixes #9673, Closes #9713 

Adds a custom HTTP parameter encoder and URL serialiser

Credits to @crizyli for original design and implementation.

### CODE REVIEW

File: src/web/services/http-request.service.ts [hunk 1]
```diff
@@ -1,10 +1,35 @@
...
 import { ResourceEndpoints } from '../types/api-endpoints';
 import { MasqueradeModeService } from './masquerade-mode.service';
 
+/**
+ * This encoder was created to support encoding and decoding of plus (+) signs.
+ *
+ * Angular will ignore the encoding for plus signs. Refer to:
+ * https://github.com/angular/angular/blob/8.0.0/packages/common/http/src/params.ts#L33
+ */
+class CustomEncoder implements HttpParameterCodec {
```
Comment (xpdavid) [inline src/web/services/http-request.service.ts:14]: - Does it get fixed in Angular 9

- Can we extend `HttpUrlEncodingCodec` and only override `encodeValue`.

File: src/web/services/http-request.service.ts [hunk 2]
```diff
@@ -1,10 +1,35 @@
...
+
+  encodeKey(key: string): string {
+    return encodeURIComponent(key);
```
Comment (wkurniawan07) [inline src/web/services/http-request.service.ts:17]: https://github.com/angular/angular/blob/b72fce8acf01639438df5ef92473e088f35e2e94/packages/common/http/src/params.ts#L81-L92

There should be a reason why they do it like this. Should just copy that and modify accordingly (if my hunch is right, remove line 88).

File: src/web/services/http-request.service.ts (hunk 1 above)
Comment (madanalogy) [reply src/web/services/http-request.service.ts:14]: I can confirm that the problem is still there. I will work on extending the codec.

Comment (madanalogy) [reply src/web/services/http-request.service.ts:14]: I just realised you're referring to a different codec than what was implemented. Are you suggesting an alternative implementation?

File: src/web/services/http-request.service.ts (hunk 2 above)
Comment (madanalogy) [reply src/web/services/http-request.service.ts:17]: Do we want to remove the slash as well? i.e. is there a list of permissible special characters for session names?

File: src/web/services/http-request.service.ts [hunk 3]
```diff
@@ -1,10 +1,35 @@
...
 import { Observable } from 'rxjs';
 import { environment } from '../environments/environment';
 import { ResourceEndpoints } from '../types/api-endpoints';
 import { MasqueradeModeService } from './masquerade-mode.service';
 
+/**
+ * This encoder was created to support encoding and decoding of plus (+) signs.
+ *
+ * Angular will ignore the encoding for plus signs. Refer to:
+ * https://github.com/angular/angular/blob/8.0.0/packages/common/http/src/params.ts#L33
```
Comment (xpdavid) [inline src/web/services/http-request.service.ts:12]: The url should be updated.

File: src/web/app/custom-url-serializer.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * Custom Url Serializer to handle plus sign related issues
+ */
+export class CustomUrlSerializer implements UrlSerializer {
+  parse(url: string): UrlTree {
+    const dus: DefaultUrlSerializer = new DefaultUrlSerializer();
+    return dus.parse(url.replace(/\+/g, '%2B'));
+  }
+
+  serialize(tree: UrlTree): any {
```
Comment (ccyccyccy) [inline src/web/app/custom-url-serializer.ts:12]: Here as well :P
```suggestion
  serialize(tree: UrlTree): string {
```

**Comment by madanalogy** [comment]
> Also, what are the roles of `CustomUrlSerializer` and `HttpParameterCodec`? Is it not sufficient to have just one of them?

I actually honestly have no idea. I tried researching on the utilisation of both the [`UrlSerializer`](https://angular.io/api/router/UrlSerializer) and [`HttpParameterCodec`](https://angular.io/api/common/http/HttpParameterCodec) and it seems that the former deals with the entire URL string while the latter is dedicated specifically towards the paramaters in a URL. I'm not sure what the difference in application is for our specific use case here. 

Would you like for me to try only have each active? And if they are both indeed fulfilling the same function, which would you rather I keep for the purposes of this fix?

**Comment by madanalogy** [comment]
> I will try out only having each active, and if they are both indeed fulfilling the same function, I will keep the encoder and drop the serializer.

@wkurniawan07 I've checked and without the URL serialiser the URL displayed in the browser does not accurately display the `+` in the feedback session name. Additionally, copying a feedback session from another when the new feedback session has a `+` in its name throws an error. These two issues do not come up with the serialiser.

Without the custom encoder the params are not properly encoded when sent to the server. I really think we need both for a comprehensive fix but I still can't figure out why exactly.

**Pull request merged**